### ISSUE TITLE
Bad interactions in JSDOMNodeJSEnv error handling on JSDom 10-12

### ISSUE THREAD

**Issue by japgolly** [scala-js/scala-js#3458]
I can't get a run of utest 0.6.4 or even 0.6.5 with the opcode change, of scalajs-react tests to complete with SJS 0.6.25. I constantly one of the following errors which kills all subsequent tests. I've had to downgrade SJS to 0.6.24

```
org.scalajs.testcommon.RPCCore$ClosedException
	at org.scalajs.testcommon.RPCCore.helpClose(RPCCore.scala:193)
	at org.scalajs.testcommon.RPCCore.close(RPCCore.scala:185)
	at org.scalajs.testadapter.ComJSEnvRPC.org$scalajs$testadapter$ComJSEnvRPC$$super$close(ComJSEnvRPC.scala:50)
	at org.scalajs.testadapter.ComJSEnvRPC$$anon$1.run(ComJSEnvRPC.scala:50)
Caused by: org.scalajs.jsenv.ComJSEnv$ComClosedException: JSCom has been closed
	at org.scalajs.jsenv.nodejs.AbstractNodeJSEnv$NodeComJSRunner$class.receive(AbstractNodeJSEnv.scala:284)
	at org.scalajs.jsenv.nodejs.JSDOMNodeJSEnv$ComDOMNodeRunner.receive(JSDOMNodeJSEnv.scala:62)
	at org.scalajs.jsenv.ComJSRunner$class.receive(ComJSRunner.scala:15)
	at org.scalajs.jsenv.nodejs.JSDOMNodeJSEnv$ComDOMNodeRunner.receive(JSDOMNodeJSEnv.scala:62)
	at org.scalajs.testadapter.ComJSEnvRPC$$anon$1.run(ComJSEnvRPC.scala:31)
Caused by: java.io.EOFException
	at java.io.DataInputStream.readInt(DataInputStream.java:392)
	at org.scalajs.jsenv.nodejs.AbstractNodeJSEnv$NodeComJSRunner$class.receive(AbstractNodeJSEnv.scala:274)
	at org.scalajs.jsenv.nodejs.JSDOMNodeJSEnv$ComDOMNodeRunner.receive(JSDOMNodeJSEnv.scala:62)
	at org.scalajs.jsenv.ComJSRunner$class.receive(ComJSRunner.scala:15)
	at org.scalajs.jsenv.nodejs.JSDOMNodeJSEnv$ComDOMNodeRunner.receive(JSDOMNodeJSEnv.scala:62)
	at org.scalajs.testadapter.ComJSEnvRPC$$anon$1.run(ComJSEnvRPC.scala:31)
```

or

```
org.scalajs.testcommon.RPCCore$ClosedException
	at org.scalajs.testcommon.RPCCore.helpClose(RPCCore.scala:193)
	at org.scalajs.testcommon.RPCCore.close(RPCCore.scala:185)
	at org.scalajs.testadapter.ComJSEnvRPC.org$scalajs$testadapter$ComJSEnvRPC$$super$close(ComJSEnvRPC.scala:50)
	at org.scalajs.testadapter.ComJSEnvRPC$$anon$1.run(ComJSEnvRPC.scala:50)
Caused by: org.scalajs.jsenv.ComJSEnv$ComClosedException: JSCom has been closed
	at org.scalajs.jsenv.nodejs.AbstractNodeJSEnv$NodeComJSRunner$class.receive(AbstractNodeJSEnv.scala:284)
	at org.scalajs.jsenv.nodejs.JSDOMNodeJSEnv$ComDOMNodeRunner.receive(JSDOMNodeJSEnv.scala:62)
	at org.scalajs.jsenv.ComJSRunner$class.receive(ComJSRunner.scala:15)
	at org.scalajs.jsenv.nodejs.JSDOMNodeJSEnv$ComDOMNodeRunner.receive(JSDOMNodeJSEnv.scala:62)
	at org.scalajs.testadapter.ComJSEnvRPC$$anon$1.run(ComJSEnvRPC.scala:31)
Caused by: java.io.EOFException
	at java.io.DataInputStream.readInt(DataInputStream.java:392)
	at org.scalajs.jsenv.nodejs.AbstractNodeJSEnv$NodeComJSRunner$class.receive(AbstractNodeJSEnv.scala:274)
	at org.scalajs.jsenv.nodejs.JSDOMNodeJSEnv$ComDOMNodeRunner.receive(JSDOMNodeJSEnv.scala:62)
	at org.scalajs.jsenv.ComJSRunner$class.receive(ComJSRunner.scala:15)
	at org.scalajs.jsenv.nodejs.JSDOMNodeJSEnv$ComDOMNodeRunner.receive(JSDOMNodeJSEnv.scala:62)
	at org.scalajs.testadapter.ComJSEnvRPC$$anon$1.run(ComJSEnvRPC.scala:31)
```

**Comment by sjrd** [comment]
Some initial investigation.

The responsible commit is 8d1938b6eaa985c933189f0eaf82edee0b3cb897, where we started using the new API of jsdom. Before this running

    > test/testOnly japgolly.scalajs.react.extra.router.RouterTest

would result in the following input, and succeed:
```
-------------------------------- Running Tests --------------------------------
[Router] Syncing to AbsUrl(http://localhost/).
[Router] Parsed Path() to Right(Root).
[Router] Action for page Root at Path() is Renderer(<function1>).
[Router] Resolved to page Root.
[Router] 
Error: Not implemented: window.scrollTo
    at module.exports (/localhome/doeraene/projects/scalajs-react/node_modules/jsdom/lib/jsdom/browser/not-implemented.js:9:17)
    at /localhome/doeraene/projects/scalajs-react/node_modules/jsdom/lib/jsdom/browser/Window.js:594:7
    at /localhome/doeraene/projects/scalajs-react/test/target/scala-2.12/test-test-fastopt.js:58962:76
    at $c_sjsr_AnonFunction0.apply__O (/localhome/doeraene/projects/scalajs-react/test/target/scala-2.12/test-test-fastopt.js:68964:23)
    at /localhome/doeraene/projects/scalajs-react/test/target/scala-2.12/test-test-fastopt.js:3309:11
    at $c_sjsr_AnonFunction0.apply__O (/localhome/doeraene/projects/scalajs-react/test/target/scala-2.12/test-test-fastopt.js:68964:23)
    at /localhome/doeraene/projects/scalajs-react/test/target/scala-2.12/test-test-fastopt.js:3678:16
    at $c_sjsr_AnonFunction0.apply__O (/localhome/doeraene/projects/scalajs-react/test/target/scala-2.12/test-test-fastopt.js:68964:23)
    at /localhome/doeraene/projects/scalajs-react/test/target/scala-2.12/test-test-fastopt.js:3678:16
    at $c_sjsr_AnonFunction0.apply__O (/localhome/doeraene/projects/scalajs-react/test/target/scala-2.12/test-test-fastopt.js:68964:23) undefined
[Router] Set route to Path(/hello).
[Router] PushState: [http://localhost//hello]
[Router] Broadcasting sync request.
[Router] Syncing to AbsUrl(http://localhost//hello).
[Router] Parsed Path(/hello) to Right(Hello).
[Router] Action for page Hello at Path(/hello) is Renderer(<function1>).
[Router] Resolved to page Hello.
[Router] 
Error: Not implemented: window.scrollTo
    at module.exports (/localhome/doeraene/projects/scalajs-react/node_modules/jsdom/lib/jsdom/browser/not-implemented.js:9:17)
    at /localhome/doeraene/projects/scalajs-react/node_modules/jsdom/lib/jsdom/browser/Window.js:594:7
    at /localhome/doeraene/projects/scalajs-react/test/target/scala-2.12/test-test-fastopt.js:58962:76
    at $c_sjsr_AnonFunction0.apply__O (/localhome/doeraene/projects/scalajs-react/test/target/scala-2.12/test-test-fastopt.js:68964:23)
    at /localhome/doeraene/projects/scalajs-react/test/target/scala-2.12/test-test-fastopt.js:3309:11
    at $c_sjsr_AnonFunction0.apply__O (/localhome/doeraene/projects/scalajs-react/test/target/scala-2.12/test-test-fastopt.js:68964:23)
    at /localhome/doeraene/projects/scalajs-react/test/target/scala-2.12/test-test-fastopt.js:5548:18
    at $c_sjsr_AnonFunction3.apply__O__O__O__O (/localhome/doeraene/projects/scalajs-react/test/target/scala-2.12/test-test-fastopt.js:69052:23)
    at .<anonymous> (/localhome/doeraene/projects/scalajs-react/test/target/scala-2.12/test-test-fastopt.js:5431:18)
    at commitLifeCycles (/tmp/1536846634384-0/react-dom.development.js:9320:24) undefined
[Router] Set route to Path().
[Router] PushState: [http://localhost/]
[Router] Broadcasting sync request.
[Router] Syncing to AbsUrl(http://localhost/).
[Router] Parsed Path() to Right(Root).
[Router] Action for page Root at Path() is Renderer(<function1>).
[Router] Resolved to page Root.
[Router] 
Error: Not implemented: window.scrollTo
    at module.exports (/localhome/doeraene/projects/scalajs-react/node_modules/jsdom/lib/jsdom/browser/not-implemented.js:9:17)
    at /localhome/doeraene/projects/scalajs-react/node_modules/jsdom/lib/jsdom/browser/Window.js:594:7
    at /localhome/doeraene/projects/scalajs-react/test/target/scala-2.12/test-test-fastopt.js:58962:76
    at $c_sjsr_AnonFunction0.apply__O (/localhome/doeraene/projects/scalajs-react/test/target/scala-2.12/test-test-fastopt.js:68964:23)
    at /localhome/doeraene/projects/scalajs-react/test/target/scala-2.12/test-test-fastopt.js:3309:11
    at $c_sjsr_AnonFunction0.apply__O (/localhome/doeraene/projects/scalajs-react/test/target/scala-2.12/test-test-fastopt.js:68964:23)
    at /localhome/doeraene/projects/scalajs-react/test/target/scala-2.12/test-test-fastopt.js:5548:18
    at $c_sjsr_AnonFunction3.apply__O__O__O__O (/localhome/doeraene/projects/scalajs-react/test/target/scala-2.12/test-test-fastopt.js:69052:23)
    at .<anonymous> (/localhome/doeraene/projects/scalajs-react/test/target/scala-2.12/test-test-fastopt.js:5431:18)
    at commitLifeCycles (/tmp/1536846634384-0/react-dom.development.js:9320:24) undefined
[Router] Set route to Path(/name/bob).
[Router] PushState: [http://localhost//name/bob]
[Router] Broadcasting sync request.
[Router] Syncing to AbsUrl(http://localhost//name/bob).
[Router] Parsed Path(/name/bob) to Right(Greet(bob)).
[Router] Action for page Greet(bob) at Path(/name/bob) is Renderer(<function1>).
[Router] Resolved to page Greet(bob).
[Router] 
Error: Not implemented: window.scrollTo
    at module.exports (/localhome/doeraene/projects/scalajs-react/node_modules/jsdom/lib/jsdom/browser/not-implemented.js:9:17)
    at /localhome/doeraene/projects/scalajs-react/node_modules/jsdom/lib/jsdom/browser/Window.js:594:7
    at /localhome/doeraene/projects/scalajs-react/test/target/scala-2.12/test-test-fastopt.js:58962:76
    at $c_sjsr_AnonFunction0.apply__O (/localhome/doeraene/projects/scalajs-react/test/target/scala-2.12/test-test-fastopt.js:68964:23)
    at /localhome/doeraene/projects/scalajs-react/test/target/scala-2.12/test-test-fastopt.js:3309:11
    at $c_sjsr_AnonFunction0.apply__O (/localhome/doeraene/projects/scalajs-react/test/target/scala-2.12/test-test-fastopt.js:68964:23)
    at /localhome/doeraene/projects/scalajs-react/test/target/scala-2.12/test-test-fastopt.js:5548:18
    at $c_sjsr_AnonFunction3.apply__O__O__O__O (/localhome/doeraene/projects/scalajs-react/test/target/scala-2.12/test-test-fastopt.js:69052:23)
    at .<anonymous> (/localhome/doeraene/projects/scalajs-react/test/target/scala-2.12/test-test-fastopt.js:5431:18)
    at commitLifeCycles (/tmp/1536846634384-0/react-dom.development.js:9320:24) undefined
[Router] Set route to Path().
[Router] PushState: [http://localhost/]
[Router] Broadcasting sync request.
[Router] Syncing to AbsUrl(http://localhost/).
[Router] Parsed Path() to Right(Root).
[Router] Action for page Root at Path() is Renderer(<function1>).
[Router] Resolved to page Root.
[Router] 
Error: Not implemented: window.scrollTo
    at module.exports (/localhome/doeraene/projects/scalajs-react/node_modules/jsdom/lib/jsdom/browser/not-implemented.js:9:17)
    at /localhome/doeraene/projects/scalajs-react/node_modules/jsdom/lib/jsdom/browser/Window.js:594:7
    at /localhome/doeraene/projects/scalajs-react/test/target/scala-2.12/test-test-fastopt.js:58962:76
    at $c_sjsr_AnonFunction0.apply__O (/localhome/doeraene/projects/scalajs-react/test/target/scala-2.12/test-test-fastopt.js:68964:23)
    at /localhome/doeraene/projects/scalajs-react/test/target/scala-2.12/test-test-fastopt.js:3309:11
    at $c_sjsr_AnonFunction0.apply__O (/localhome/doeraene/projects/scalajs-react/test/target/scala-2.12/test-test-fastopt.js:68964:23)
    at /localhome/doeraene/projects/scalajs-react/test/target/scala-2.12/test-test-fastopt.js:5548:18
    at $c_sjsr_AnonFunction3.apply__O__O__O__O (/localhome/doeraene/projects/scalajs-react/test/target/scala-2.12/test-test-fastopt.js:69052:23)
    at .<anonymous> (/localhome/doeraene/projects/scalajs-react/test/target/scala-2.12/test-test-fastopt.js:5431:18)
    at commitLifeCycles (/tmp/1536846634384-0/react-dom.development.js:9320:24) undefined
Warning: unmountComponentAtNode(): The node you're attempting to unmount was rendered by React and is not a top-level container. You may have accidentally passed in a React root node instead of its container.
+ japgolly.scalajs.react.extra.router.RouterTest.sim 74ms  
+ japgolly.scalajs.react.extra.router.RouterTest.pure.urlParsing.root 0ms  
+ japgolly.scalajs.react.extra.router.RouterTest.pure.urlParsing.tslash 0ms  
+ japgolly.scalajs.react.extra.router.RouterTest.pure.urlParsing.path 0ms  
[SimHistory] Parsed Path(/what) to Left(RedirectToPage(Root,Replace)).
[SimHistory] Redirecting to AbsUrl(http://www.yaya.com/blah) via Replace.
[SimHistory] Parsed Path() to Right(Root).
[SimHistory] Action for page Root at Path() is Renderer(<function1>).
+ japgolly.scalajs.react.extra.router.RouterTest.pure.syncToUrl.notFound_redirect 1ms  
[SimHistory] Wrong base: AbsUrl(https://www.google.com) is outside of AbsUrl(http://www.yaya.com/blah).
[SimHistory] Redirecting to AbsUrl(http://www.yaya.com/blah) via Push.
[SimHistory] Parsed Path() to Right(Root).
[SimHistory] Action for page Root at Path() is Renderer(<function1>).
+ japgolly.scalajs.react.extra.router.RouterTest.pure.syncToUrl.badbase 0ms  
[SimHistory] Parsed Path(/) to Left(RedirectToPath(Path(),Replace)).
[SimHistory] Redirecting to AbsUrl(http://www.yaya.com/blah) via Replace.
[SimHistory] Parsed Path() to Right(Root).
[SimHistory] Action for page Root at Path() is Renderer(<function1>).
+ japgolly.scalajs.react.extra.router.RouterTest.pure.syncToUrl.tslash_root 0ms  
[SimHistory] Parsed Path(/hello/) to Left(RedirectToPath(Path(/hello),Replace)).
[SimHistory] Redirecting to AbsUrl(http://www.yaya.com/blah/hello) via Replace.
[SimHistory] Parsed Path(/hello) to Right(Hello).
[SimHistory] Action for page Hello at Path(/hello) is Renderer(<function1>).
+ japgolly.scalajs.react.extra.router.RouterTest.pure.syncToUrl.tslash_path 0ms  
[info] Tests: 8, Passed: 8, Failed: 0
[success] Total time: 70 s, completed Sep 13, 2018 3:50:37 PM
```
A similar output can be obtained even with 0.6.25, by forcing `jsdom@9.12.0` (in which case we fall back to the old API of jsdom).

After the mentioned commit and with jsdom 11 or 12 (and I guess 10), we obtain the following instead:
```
-------------------------------- Running Tests --------------------------------
[Router] Syncing to AbsUrl(http://localhost/).
[Router] Parsed Path() to Right(Root).
[Router] Action for page Root at Path() is Renderer(<function1>).
[Router] Resolved to page Root.
[Router] 
{ Error: Not implemented: window.scrollTo
    at module.exports (/localhome/doeraene/projects/scalajs-react/node_modules/jsdom/lib/jsdom/browser/not-implemented.js:9:17)
    at /localhome/doeraene/projects/scalajs-react/node_modules/jsdom/lib/jsdom/browser/Window.js:594:7
    at file:///localhome/doeraene/projects/scalajs-react/test/target/scala-2.12/test-test-fastopt.js:58962:76
    at $c_sjsr_AnonFunction0.apply__O (file:///localhome/doeraene/projects/scalajs-react/test/target/scala-2.12/test-test-fastopt.js:68964:23)
    at file:///localhome/doeraene/projects/scalajs-react/test/target/scala-2.12/test-test-fastopt.js:3309:11
    at $c_sjsr_AnonFunction0.apply__O (file:///localhome/doeraene/projects/scalajs-react/test/target/scala-2.12/test-test-fastopt.js:68964:23)
    at file:///localhome/doeraene/projects/scalajs-react/test/target/scala-2.12/test-test-fastopt.js:3678:16
    at $c_sjsr_AnonFunction0.apply__O (file:///localhome/doeraene/projects/scalajs-react/test/target/scala-2.12/test-test-fastopt.js:68964:23)
    at file:///localhome/doeraene/projects/scalajs-react/test/target/scala-2.12/test-test-fastopt.js:3678:16
    at $c_sjsr_AnonFunction0.apply__O (file:///localhome/doeraene/projects/scalajs-react/test/target/scala-2.12/test-test-fastopt.js:68964:23) type: 'not implemented' }
[trace] Stack trace suppressed: run last test/test:testOnly for the full output.
[error] Could not run test japgolly.scalajs.react.extra.router.RouterTest: org.scalajs.testcommon.RPCCore$ClosedException
[trace] Stack trace suppressed: run last test/test:testOnly for the full output.
[error] java.lang.NullPointerException
[trace] Stack trace suppressed: run last test/test:testOnly for the full output.
[error] (test/test:testOnly) org.scalajs.testcommon.RPCCore$ClosedException
[error] Total time: 71 s, completed Sep 13, 2018 3:55:58 PM
```
which results in an error.

I am not sure whether the error that appears is a *progression* or a *regression*. Even with Scala.js 0.6.24, we can see that a JavaScript stack trace is output, for the error `Error: Not implemented: window.scrollTo`. However, in Scala.js 0.6.24, that error is *ignored* and testing continues. In 0.6.25, that error kills the process to make the tests fail.

I *think* the new behavior is actually the desired behavior, and that it discovers an actual bug in scalajs-react that was previously overlooked by ignoring the `Not implemented: window.scrollTo` error.

@japgolly WDYT?

**Comment by sjrd** [comment]
Relevant commit in jsdom: https://github.com/jsdom/jsdom/commit/456a678179ca8adc5a4a33d2f46096d9e978e026

Maybe we should only fail the process if a `jsdomError` whose `type` is `"unhandled exception"` is thrown?

**Comment by japgolly** [comment]
Something is definitely up with the new RPC protocol and/or its handling in 0.6.25. Here's a different example. I've kind of minimised it to a single scalajs-react test case that passes/fails depending on the SJS version (and no other differences).

* There are two commits A and B in scalajs-react: d50588214ec67da520d4ab85263360b994def10e and eb77b2888815472d0731830340193f9fb02cf1f3 respectively
* A is the parent of B and you can see the only difference is the SJS version
* Both A and B use utest 0.6.5
* Both A and B use Scala 2.12.6
* Both A and B use React 16.5.1
* Both A and B use succeed on my computer
* A consistently fails on Travis CI with RPCCore exception, B passes on Travis CI
* A is on SJS 0.6.25, B is on SJS 0.6.24

The run just the unit test in question from SBT:
```
test/testOnly -- japgolly.scalajs.react.core.ScalaComponentPTest.lifecycle1
```

What the unit test does is render an outer component with an inner component, then pass a `null` to the inner component deliberately to make it crash. Because the outer component provides a React callback `componentDidCatch`, it is expected that the exception be caught. The test then asserts that the outer components renders the caught error message.

As part of my stdout-debugging it can be observed that with SJS 0.6.25, the error isn't caught by React somehow. Even wrapping the entire test in a `try/catch` where I catch all exceptions, not just `NonFatal` ones, the error escapes somehow and breaks the RPC protocol.

Here is the relevant Travis output for [(A / SJS 0.6.25 / failure)](https://travis-ci.org/japgolly/scalajs-react/jobs/428930835#L849-L919) and [(B / SJS 0.6.24 / success)](https://travis-ci.org/japgolly/scalajs-react/jobs/428939543#L849-L919).

And here is the corresponding code: https://github.com/japgolly/scalajs-react/blob/eb77b2888815472d0731830340193f9fb02cf1f3/test/src/test/scala/japgolly/scalajs/react/core/ScalaComponentTest.scala#L180-L199

**Comment by sjrd** [comment]
Thanks for the detailed analysis. I'll try to look into it in the coming days.

Is the error also observable on 0.6.25 if you force `npm install jsdom@9.12.0`?

**Comment by japgolly** [comment]
Well this is very surprising to me! I didn't expect this to bare fruit but I made [this commit](https://github.com/japgolly/scalajs-react/commit/ffc405dd8125a5a1cc3050194a47bd771fb343da) and sent it off to Travis and [it worked](https://travis-ci.org/japgolly/scalajs-react/jobs/428950375#L849-L920)! I don't see how jsdom relates to this specific unit test; it's not like the `window.scrollTo not implemented` error. But yeah, well done! jsdom is relevant.

Other info, some of which I forgot to mention above:
* Using jsdom 9.12.0 locally with 0.6.25 passed too
* Both locally and in Travis, here and above, I've been using Node 10 and jsdom@11

**Comment by sjrd** [comment]
It's again 8d1938b6eaa985c933189f0eaf82edee0b3cb897. With jsdom 9.x we use a *completely different implementation* of the `JSDOMNodeJSEnv`. We used to also use the same implem with the legacy API in 10 and 11. The commit above switched to using the new API starting from jsdom 10+, to support jsdom 12. But apparently our new implementation doesn't have exactly the same reaction to internal exceptions, and does something wrong. We'll have to figure out what exactly.

**Comment by sjrd** [comment]
After some more investigation on your reproduction with `lifecycle1`, it seems that there is a bad interaction between [a hack that React itself uses in Dev mode](https://github.com/facebook/react/blob/17e703cb9692c26795c794a54e2239f603ee090d/packages/shared/invokeGuardedCallbackImpl.js#L32-L49), our way of handling uncaught errors, and jsdom's own handling of uncaught errors.

From our point of view, an uncaught error should terminate the VM. Otherwise, exceptions in scripts won't be reported as failed executions (meaning a CI would be green when it was supposed to fail).

From jsdom's point of view, uncaught exceptions are sent to a special `jsdomError` event handler in the virtual console. This should only happen at the top-level of jsdom's handling, which is outside the testing framework's `try catch` which is supposed to catch errors, so in general this is fine.

But now enters React's hack: it forces *synchronous* dispatch of an error using relatively "low-level" APIs of the `Window`, which turns *caught* exceptions to appear to jsdom as if they were *uncaught*. Now jsdom sends those to our handler which kills the VM.

And there you have it: React's hack turns a caught exception into uncaught (to "integrate better with the browsers' DevTools, or at least their *default settings*, sic!), which causes us to believe it is a hard error which must be reported as a VM kill. The VM kill subsequently "breaks the RPC" (in fact it is the correct behavior of the RPC to report hard errors when VM is killed).

At this point I'm not sure what we can do about it. What we do in jsdom is consistent with the non-jsdom Node.js environment: an uncaught error is a hard failure, whether thrown from the script top-level or from a timeout/etc. Technically we could hack our jsdom env into only considering script top-level uncaught exceptions to be hard failures, which might solve this specific bad interact. In fact I suspect that this was what the previous implementation did, but that was incorrect wrt. our `JSEnv` model.

**Comment by japgolly** [comment]
Wow that's some very awesome (and seemingly-time-consuming) debugging. Unfortunately though, I don't think it's going to be enough to fully explain the issue :disappointed: Here's a new data point:

In another library of mine called "Scala Test-State" I got the [same exception in Travis](https://travis-ci.org/japgolly/test-state/jobs/437860568#L2167-L2180). It's using Scala.JS 0.6.25, μtest 0.6.5 and [.travis.yml is configured to use jsdom 9.12.0](https://github.com/japgolly/test-state/blob/88c9eb1892127d9994fd0e0c90f9a565a832e75b/.travis.yml#L19-L24). What's particularly interesting about this is that the exception occurred in `coreJS/test` where all the tests are shared with the JVM and don't touch any DOM. Also interesting is that I've only seen this exception for test-state in Travis and not locally.

**Comment by japgolly** [comment]
...and yet test-state's `coreJS` seems to be fine with [jsdom 11](https://travis-ci.org/japgolly/test-state/jobs/437873548) & [jsdom 12](https://travis-ci.org/japgolly/test-state/jobs/437873009). So even though jsdom isn't being used in the `coreJS` tests, the new behaviour it enables in Scala.JS seems to work where as the old has some problems with error-handling in general.

**Comment by sjrd** [comment]
@gutee Do you have a reproduction? I.e., complete code and precise steps to reproduce that issue?

**Comment by gutee** [comment]
I couldn't reproduce this locally, but it happens randomly in approximately 20% of the builds in Jenkins. We are using scala 2.12.6, sbt 0.13.16 and scalatest 3.0.0.
Using NodeJSEnv we run the tests in parallel, JVM is successful, but JS tests are interrupted by this.

**Comment by sjrd** [comment]
> Wow that's some very awesome (and seemingly-time-consuming) debugging. Unfortunately though, I don't think it's going to be enough to fully explain the issue :disappointed: Here's a new data point:

Looking more closely at this, I don't think the new data point invalidates the previous reasoning, because it happens with jsdom 9.x and *not* jsdom 11/12. So it seems completely unrelated.

> In another library of mine called "Scala Test-State" I got the [same exception in Travis](https://travis-ci.org/japgolly/test-state/jobs/437860568#L2167-L2180). It's using Scala.JS 0.6.25, μtest 0.6.5 and [.travis.yml is configured to use jsdom 9.12.0](https://github.com/japgolly/test-state/blob/88c9eb1892127d9994fd0e0c90f9a565a832e75b/.travis.yml#L19-L24). What's particularly interesting about this is that the exception occurred in `coreJS/test` where all the tests are shared with the JVM and don't touch any DOM. Also interesting is that I've only seen this exception for test-state in Travis and not locally.
>
> ...and yet test-state's `coreJS` seems to be fine with [jsdom 11](https://travis-ci.org/japgolly/test-state/jobs/437873548) & [jsdom 12](https://travis-ci.org/japgolly/test-state/jobs/437873009). So even though jsdom isn't being used in the `coreJS` tests, the new behaviour it enables in Scala.JS seems to work where as the old has some problems with error-handling in general.

This suggests that the new behavior is at least better for some things. So even if we reverted wholesale to the previous behavior (somehow while making it work on jsdom 12+), we would still regress on that path.

I am increasingly uncomfortable about this issue. I don't know what to do or try next.

**Comment by gzm0** [comment]
Assigning this tentatively to the 0.6 branch, as JSDomNodeJSEnv is in a separate repo in the 1.0 branch.

**Comment by sjrd** [comment]
Fixed in f4ecc893010638fc48aa61e7ec328a255c4838f4.
